### PR TITLE
Add adversarial training example

### PR DESCRIPTION
**Pull request by mmhamdy**
Migrating [jaxopt's example on adversarial training](https://jaxopt.github.io/stable/notebooks/deep_learning/adversarial_training.html) to optax, addressing issue #677

### CODE REVIEW

**Comment by fabianp** [comment]
Thanks for the pull request @mmhamdy ! Are you still working on it or you'd like me to review it?

**Comment by mmhamdy** [comment]
Hi @fabianp, glad to help! you can review it.

File: examples/adversarial_training.ipynb [hunk 1]
```diff
@@ -0,0 +1,1052 @@
...
+        "import collections\n",
+        "import operator\n",
+        "\n",
+        "# activate TPUs if available\n",
+        "try:\n",
+        "    import jax.tools.colab_tpu\n",
+        "    jax.tools.colab_tpu.setup_tpu()\n",
+        "except (KeyError, RuntimeError):\n",
+        "    print(\"TPU not found, continuing without it.\")\n",
+        "\n",
```
Comment (fabianp) [inline examples/adversarial_training.ipynb:97]: you can remove the try/except block, it's no longer needed

File: examples/adversarial_training.ipynb [hunk 2]
```diff
@@ -0,0 +1,1052 @@
...
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "cib2jXJDQ4Nz"
+      },
+      "source": [
+        "# Adversarial training\n",
+        "\n",
+        "\n",
+        "[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/google/jaxopt/blob/main/docs/notebooks/deep_learning/adversarial_training.ipynb)\n",
```
Comment (fabianp) [inline examples/adversarial_training.ipynb:33]: please update this URL (to reflect the optax repo instead of the jaxopt one)

File: examples/adversarial_training.ipynb [hunk 3]
```diff
@@ -0,0 +1,1052 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 4,
+      "metadata": {
+        "id": "HLuuYP6LQ4N6"
+      },
+      "outputs": [],
+      "source": [
+        "Flags = collections.namedtuple(\n",
```
Comment (fabianp) [inline examples/adversarial_training.ipynb:154]: please use the new flags format we use for example in https://optax.readthedocs.io/en/latest/_collections/examples/cifar10_resnet.html (the cell that starts with @markdown). This format doesn't require to define the Flags namedtuple and instead stores everything as global variables

File: examples/adversarial_training.ipynb [hunk 4]
```diff
@@ -0,0 +1,1052 @@
...
+        "net = CNN(num_classes)\n",
+        "\n",
+        "@jax.jit\n",
+        "def accuracy(params, data):\n",
+        "  inputs, labels = data\n",
+        "  logits = net.apply({\"params\": params}, inputs)\n",
+        "  return jnp.mean(jnp.argmax(logits, axis=-1) == labels)\n",
+        "\n",
+        "@jax.vmap\n",
+        "def logistic_loss(label: int, logits: jnp.ndarray) -> float:\n",
```
Comment (fabianp) [inline examples/adversarial_training.ipynb:292]: please use one of optax's predefined losses (in this case I think it's https://optax.readthedocs.io/en/latest/api.html#optax.softmax_cross_entropy_with_integer_labels )

File: examples/adversarial_training.ipynb [hunk 5]
```diff
@@ -0,0 +1,1052 @@
...
+        "  \"\"\"Multiclass logistic loss.\"\"\"\n",
+        "  logits = jnp.asarray(logits)\n",
+        "  # To avoid roundoff error, subtract target inside logsumexp.\n",
+        "  # logsumexp(logits) - logits[y] = logsumexp(logits - logits[y])\n",
+        "  logits = (logits - logits[label]).at[label].set(0.0)\n",
+        "  return jax.scipy.special.logsumexp(logits)\n",
+        "\n",
+        "def tree_l2_norm(tree_x, squared=False):\n",
```
Comment (fabianp) [inline examples/adversarial_training.ipynb:300]: Use `optax.tree_utils.tree_l2_norm` instead (see source code here https://github.com/google-deepmind/optax/blob/master/optax/tree_utils/_tree_math.py#L150 )

File: examples/adversarial_training.ipynb [hunk 6]
```diff
@@ -0,0 +1,1052 @@
...
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "Dataset mnist downloaded and prepared to /root/tensorflow_datasets/mnist/3.0.1. Subsequent calls will reuse this data.\n"
+          ]
+        }
+      ],
+      "source": [
+        "train_ds, ds_info = load_dataset(\"train\", is_training=True,\n",
```
Comment (fabianp) [inline examples/adversarial_training.ipynb:273]: please use a simpler loader like the one in https://optax.readthedocs.io/en/latest/_collections/examples/mlp_mnist.html that doesn't define its own load_dataset function. I believe also this might help fix the test errors we're getting

**Comment by fabianp** [comment]
Thanks again for the contribution. Besides the comments that I left in the file, it would be important to:

1. Fix the test errors (see https://github.com/google-deepmind/optax/actions/runs/7300424034/job/19895029972?pr=682)
2. Add this example to the gallery (https://github.com/google-deepmind/optax/blob/master/docs/gallery.rst)

**Comment by mmhamdy** [comment]
Thanks, @fabianp. Sounds great! Working on them.

File: examples/adversarial_training.ipynb (hunk 5 above)
Comment (mmhamdy) [reply examples/adversarial_training.ipynb:300]: This is only available in the dev version.

Comment (fabianp) [reply examples/adversarial_training.ipynb:300]: it's fine if the examples only run with the dev version. Plus we'll be releasing a new version very soon

Comment (mmhamdy) [reply examples/adversarial_training.ipynb:300]: Great! will add a comment indicating why the dev version was used.

**Comment by fabianp** [comment]
No need. As I said we'll release a new version as soon as possible

**Comment by fabianp** [comment]
haha yeah, removing the TF dependency is probably more challenging! Should I take another look at this PR or wait a bit more?

**Comment by mmhamdy** [comment]
I guess you can take another look. Happy to address any other comments.

File: docs/gallery.rst [hunk 7]
```diff
@@ -172,6 +172,23 @@ Examples that make use of the :ref:`contrib` module.
 
     <div class="sphx-glr-thumbnails">
 
+.. raw:: html
+
+    <div class="sphx-glr-thumbcontainer" tooltip="Adversarial training of CNN on MNIST.">
+
+.. only:: html
+
+  .. image:: /images/favicon.svg
```
Comment (fabianp) [inline docs/gallery.rst:181]: It would be great if you can use an example-specific image, such as this one instead of the generic favicon.svg one (see for example how it's done in the "Optimistic Gradient Descent in a Bilinear Min-Max Problem" example)

image

Comment (fabianp) [reply docs/gallery.rst:181]: Having a nice looking thumbnail goes a long way in making the gallery pretty and the example easy to find :-)

Comment (mmhamdy) [reply docs/gallery.rst:181]: Happy New Year ❄🎉. Sure, I agree. Will add this image as a thumbnail. I was thinking maybe we could create a thumbnail template for the example gallery. Tell me what you think. If you agree, I can work on some sample designs.

**Comment by fabianp** [comment]
What do mean by thumbnail template? Like a default icon that's now the
favicon one?

**Comment by mmhamdy** [comment]
Like a unified template with a fixed layout where you insert an example-specific image. The rest of the layout may contain, for example, title and optax logo.

**Comment by fabianp** [comment]
I wonder if that would  make it harder for contributors to add new images if it requires using some image program like gimp (but perhaps I'm not getting your idea) ?

In any case let's leave that for a different issue/pull request? I wouldn't want to delay merging this one that's almost ready :-)

**Comment by mmhamdy** [comment]
You're right, it would certainly require some extra work from contributors. A plain thumbnail is practical and easier to add.
